query/plugins/org.eclipse.viatra.query.patternlanguage.emf.ui/src/org/eclipse/viatra/query/patternlanguage/emf/ui/util/JavaProjectClassLoaderProvider.java

```diff
@@ -42,25 +43,40 @@ public class JavaProjectClassLoaderProvider extends SimpleClassLoaderProvider im
 
     @Inject
     private IWorkspaceRoot root;
+    
+    @Inject
+    private IResourceScopeCache cache;
 
     @Override
     public ClassLoader getClassLoader(EObject ctx) {
-        try {
-            IFile file = getIFile(ctx);
-            ClassLoader l;
-            if (file != null && file.exists()) {
-                l = getClassLoader(file);
-                if (l == null) {
-                    throw new ViatraQueryException(String.format("No classloader found for context object %s.", ctx), "No classloader found.");
+        if (null == ctx || null == ctx.eResource()) {
+            return super.getClassLoader(ctx);
+        }
+        // Caching is necessary not just for performance, 
+        // but also in order to force the same source class URI to be loaded by the same ClassLoader 
+        // (hence resulting in the same Class<T> object) 
+        // for all parts of a Pattern.
+        // 
+        // Caveat: the classpath may change without anybody touching this resource. 
+        // However, a simple editing of this VQL resource, or a clean&build, will fix the stale classpath problem, while there is no simple fix for the alternative. 
```

Comment:
There is a narrow window of error if classpath is stale: the user may e.g. change some classes in source files without touching the vql, and Wuery Explorer would still work with the old versions of the classes. Note that this cannot happen to classes that are compiled and part of the runtime Eclipse. 

On  the other hand, there is a huge window of error if the classloader is never reused. For instance, the test patterns added with this PR would break in a runtime Eclipse UI (but frustratingly would still pass automated tests that use the Simple classload mechanism). See, the Enum BackendRequirement loaded as a java constant expression or eval and the one decorating the header as a parameter type filter will be different Class<> objects from different classloaders, and hence the type filtering constraint will always fail.

Reply:
I have tested the behaviour with a debugger, and was unable to produce a state of staleness. Apparently, Xtext/Xbase will evict the cache even if the change only affects other files (I experimented with breaking the superclass relationship between two Java classes, neither of them directly mentioned in the `.vql`). Whether Xtext will always do this eviction when necessary, I cannot tell. But we will be no worse than Xbase itself. 

I currently see no reason to reject the application of this cache, and am looking forward to both correctness and performance improvements enabled by this PR.

Reply:
Understood, but please give me a bit of time to think about this approach, I am still wary of this. If I cannot find anything better before we are finalizing version 2.9, I promise to merge this change as it is now, but please give me a few days to think about other possible approaches.

Reply:
I don't really see any real alternative, so I am prepared to merge this approach.